compiler/rustc_infer/src/infer/opaque_types.rs

```diff
@@ -490,6 +499,19 @@ where
                 substs.as_generator().yield_ty().visit_with(self);
                 substs.as_generator().resume_ty().visit_with(self);
             }
+            ty::Opaque(def_id, substs) => {
+                // Don't go into the same opaque type again.
+                if ty != self.skip {
+                    let skip = std::mem::replace(&mut self.skip, ty);
+                    // Only lifetimes in the bounds are relevant.
+                    // This is sound because we only permit lifetimes from the bounds to appear in the hidden type.
+                    let predicates = self.tcx.bound_explicit_item_bounds(def_id);
+                    for (predicate, _) in predicates.0 {
```

Comment:
```suggestion
                    for (predicate, _) in predicates.skip_binder() {
```
you may have to add `fn skip_binder` to `EarlyBinder` for that